components/places/src/bookmark_sync/store.rs

```diff
@@ -0,0 +1,1233 @@
+    }
+}
+
+impl<'a> BookmarksStore<'a> {
+    fn stage_incoming(
+        &self,
+        inbound: IncomingChangeset,
+        incoming_telemetry: &mut telemetry::EngineIncoming,
+    ) -> Result<ServerTimestamp> {
+        let timestamp = inbound.timestamp;
+        let mut tx = self
+            .db
+            .time_chunked_transaction(Duration::from_millis(1000))?;
+
+        let applicator = IncomingApplicator::new(&self.db);
+
+        for incoming in inbound.changes {
+            applicator.apply_payload(incoming.0, timestamp)?;
+            incoming_telemetry.applied(1);
+            tx.maybe_commit()?;
+        }
+        tx.commit()?;
+        Ok(timestamp)
+    }
+
+    fn has_changes(&self) -> Result<bool> {
+        // In the first subquery, we check incoming items with needsMerge = true
+        // except the tombstones who don't correspond to any local bookmark because
+        // we don't store them yet, hence never "merged" (see bug 1343103).
+        let sql = format!(
+            "
+            SELECT
+              EXISTS (
+               SELECT 1
+               FROM moz_bookmarks_synced v
+               LEFT JOIN moz_bookmarks b ON v.guid = b.guid
+               WHERE v.needsMerge AND
+               (NOT v.isDeleted OR b.guid NOT NULL)
+              ) OR EXISTS (
+               WITH RECURSIVE
+               {}
+               SELECT 1
+               FROM localItems
+               WHERE syncChangeCounter > 0
+              ) OR EXISTS (
+               SELECT 1
+               FROM moz_bookmarks_deleted
+              )
+              AS hasChanges
+        ",
+            *LOCAL_ITEMS_SQL_FRAGMENT
+        );
+        Ok(self
+            .db
+            .try_query_row(
+                &sql,
+                &[],
+                |row| -> rusqlite::Result<_> { Ok(row.get_checked::<_, bool>(0)?) },
+                false,
+            )?
+            .unwrap_or(false))
+    }
+
+    /// Builds a temporary table with the merge states of all nodes in the merged
+    /// tree, then updates the local tree to match the merged tree.
+    ///
+    /// Conceptually, we examine the merge state of each item, and either leave the
+    /// item unchanged, upload the local side, apply the remote side, or apply and
+    /// then reupload the remote side with a new structure.
+    fn update_local_items<'t>(
+        &self,
+        descendants: Vec<MergedDescendant<'t>>,
+        deletions: Vec<Deletion>,
+    ) -> Result<()> {
+        // First, insert rows for all merged descendants.
+        sql_support::each_sized_chunk(
+            &descendants,
+            sql_support::default_max_variable_number() / 4,
+            |chunk, _| -> Result<()> {
+                // We can't avoid allocating here, since we're binding four
+                // parameters per descendant. Rust's `SliceConcatExt::concat`
+                // is semantically equivalent, but requires a second allocation,
+                // which we _can_ avoid.
```

Comment:
...which we _can_ avoid by writing this out.

@thomcc, can you think of any way to tighten this up?